x/confidentialtransfers/keeper/msg_server.go

```diff
@@ -2,48 +2,450 @@ package keeper
+	}
+
+	// Verify that the remaining balance sent by the user matches the remaining balance calculated by the server.
+	teg := elgamal.NewTwistedElgamal()
+	remainingBalanceCalculated, err := teg.SubScalar(account.AvailableBalance, instruction.Amount)
+	if err != nil {
+		return nil, sdkerrors.Wrap(sdkerrors.ErrInvalidRequest, "Error subtracting amount")
+	}
+
+	verified = zkproofs.VerifyCiphertextCommitmentEquality(instruction.Proofs.RemainingBalanceEqualityProof, &account.PublicKey, remainingBalanceCalculated, &instruction.RemainingBalanceCommitment.C)
+	if !verified {
+		return nil, sdkerrors.Wrap(sdkerrors.ErrInvalidRequest, "Ciphertext Commitment equality verification failed")
+	}
+
+	// Update the account state
+	account.DecryptableAvailableBalance = instruction.DecryptableBalance
+	account.AvailableBalance = remainingBalanceCalculated
+
+	// Save the account state
+	m.Keeper.SetAccount(ctx, address, req.Denom, account)
+
+	// Return the tokens to the sender
+	coins := sdk.NewCoins(sdk.NewCoin(instruction.Denom, sdk.NewIntFromUint64(instruction.Amount)))
+	if err := m.Keeper.SendTokens(ctx, address, coins); err != nil {
+		return nil, sdkerrors.Wrapf(sdkerrors.ErrInsufficientFunds, "insufficient funds to withdraw %d %s", req.Amount, req.Denom)
+	}
+
+	// Emit any required events
+	ctx.EventManager().EmitEvents(sdk.Events{
+		sdk.NewEvent(
+			types.EventTypeWithdraw,
+			sdk.NewAttribute(types.AttributeDenom, instruction.Denom),
+			sdk.NewAttribute(types.AttributeAddress, instruction.FromAddress),
+			sdk.NewAttribute(sdk.AttributeKeyAmount, sdk.NewCoin(instruction.Denom, sdk.NewIntFromUint64(instruction.Amount)).String()),
+		),
+	})
+
+	return &types.MsgWithdrawResponse{}, nil
 }
 
-func (m msgServer) CloseAccount(ctx context.Context, account *types.MsgCloseAccount) (*types.MsgCloseAccountResponse, error) {
-	//TODO implement me
-	panic("implement me")
+func (m msgServer) ApplyPendingBalance(goCtx context.Context, req *types.MsgApplyPendingBalance) (*types.MsgApplyPendingBalanceResponse, error) {
+	ctx := sdk.UnwrapSDKContext(goCtx)
+
+	// Check if the account exists
+	address, err := sdk.AccAddressFromBech32(req.Address)
+	if err != nil {
+		return nil, sdkerrors.Wrap(sdkerrors.ErrInvalidRequest, "Invalid Address")
+	}
+
+	account, exists := m.Keeper.GetAccount(ctx, address, req.Denom)
+	if !exists {
+		return nil, sdkerrors.Wrap(sdkerrors.ErrInvalidRequest, "Account does not exist")
+	}
+
+	if account.PendingBalanceCreditCounter == 0 {
+		return nil, sdkerrors.Wrap(sdkerrors.ErrInvalidRequest, "No pending balance to apply")
+	}
+
+	// Calculate updated balances
+	teg := elgamal.NewTwistedElgamal()
+	newAvailableBalance, err := teg.AddWithLoHi(account.AvailableBalance, account.PendingBalanceLo, account.PendingBalanceHi)
+	if err != nil {
+		return nil, sdkerrors.Wrap(sdkerrors.ErrInvalidRequest, "Error summing balances")
+	}
+
+	zeroCiphertextLo, err := elgamal.SubtractCiphertext(account.PendingBalanceLo, account.PendingBalanceLo)
```

Comment:
We use subtraction here since Encryption is non deterministic (since it generates randomness), but subtraction is a deterministic operation.

From a confidentiality pov, It somewhat makes sense to store nil here since the pending balance counter will be set to 0 instead so everyone knows there are no pending balances in the account. 
But from a implementation perspective I think its much easier to deal with accounts if we just assume the pending balances and available balances are always a ciphertext (don't have to check for nil, write special edge cases for when pending balance is nil etc, we can just perform arithmetic on the ciphertext)